libs/wire-subsystems/src/Wire/IndexedUserStore/ElasticSearch.hs

```diff
@@ -91,31 +93,45 @@ upsertImpl cfg docId userDoc versioning = do
   where
     indexDoc :: ES.IndexName -> ES.BH (Sem r) ()
     indexDoc idx = do
-      r <- ES.indexDocument idx mappingName settings userDoc docId
+      r <-
+        hoistBH (embed @IO) $
+          ES.performBHRequest . fmap fst . ES.keepBHResponse $
+            ESR.indexDocument idx settings userDoc docId
       unless (ES.isSuccess r || ES.isVersionConflict r) $ do
         lift $ Metrics.incCounter indexUpdateErrorCounter
-        res <- liftIO $ ES.parseEsResponse r
-        liftIO . throwIO . IndexUpdateError . either id id $ res
+        liftIO . throwIO . IndexUpdateError $ parseESError r
       lift $ Metrics.incCounter indexUpdateSuccessCounter
 
     settings = ES.defaultIndexDocumentSettings {ES.idsVersionControl = versioning}
 
+hoistBH :: (forall x. m x -> n x) -> ES.BH m a -> ES.BH n a
+hoistBH nat (ES.BH action) = ES.BH $ hoistReaderT (hoistExceptT nat) action
+
+hoistReaderT :: (forall x. m x -> n x) -> ReaderT r m a -> ReaderT r n a
+hoistReaderT nat (ReaderT f) = ReaderT $ \r -> nat (f r)
+
+-- Hoist a natural transformation from m to n through ExceptT
+hoistExceptT :: (forall x. m x -> n x) -> ExceptT e m a -> ExceptT e n a
+hoistExceptT nat (ExceptT ema) = ExceptT (nat ema)
+
+castResponse :: forall context1 val1 context2 val2. BHResponse context1 val1 -> BHResponse context2 val2
+castResponse BHResponse {..} = BHResponse {..}
+
 updateTeamSearchVisibilityInboundImpl :: forall r. (Member (Embed IO) r) => IndexedUserStoreConfig -> TeamId -> SearchVisibilityInbound -> Sem r ()
 updateTeamSearchVisibilityInboundImpl cfg tid vis =
   void $ runInBothES cfg updateAllDocs
   where
     updateAllDocs :: ES.IndexName -> ES.BH (Sem r) ()
     updateAllDocs idx = do
-      r <- ES.updateByQuery idx query (Just script)
+      r <- hoistBH (embed @IO) $ ES.performBHRequest . fmap fst . ES.keepBHResponse $ ESR.updateByQuery @Value idx query (Just script)
```

Comment:
This handling of version conflicts is probably broken here. See https://github.com/bitemyapp/bloodhound/issues/295